File: inventory_tools/inventory_tools/report/specification/specification.py
# Copyright (c) 2023, AgriTheory and contributors
# For license information, please see license.txt

from itertools import groupby

import frappe


def execute(filters=None):
	specification = frappe.get_doc("Specification", filters.specification)
	return get_columns(filters, specification), get_data(filters, specification)


def get_data(filters, specification):
	doctypes = [row.applied_on for row in specification.attributes]
	attributes = [row.attribute_name for row in specification.attributes]
	fields = {row.attribute_name: row.field for row in specification.attributes if row.field}
	spec_values = frappe.get_all(
		"Specification Value",
		{"reference_doctype": ["in", doctypes], "attribute": ["in", attributes]},
		["reference_doctype", "reference_name", "attribute", "value", "name"],
		order_by="reference_name",
	)
	data = []
	for ref, d in groupby(spec_values, key=lambda x: x.get("reference_name")):
		reference_rows = sorted(
			sorted(list(d), key=lambda x: x.get("value")), key=lambda x: x.get("attribute")
		)
		data.append(
			{
				"reference_name": frappe.bold(ref),
				"indent": 0,
			}
		)
		for row in reference_rows:
			if row.attribute in fields:
				row.field = fields[row.attribute]
			row.indent = 1
			data.append(row)
	return data


def get_columns(filters, specification):
	return [
		{
			"label": f"{specification.dt} - {specification.apply_on}"
			if specification.apply_on
			else specification.dt,
			"fieldname": "reference_name",
			"fieldtype": "Link",
			"options": "Doctype",
			"width": "250px",
		},
		{
			"label": "DocType",
			"fieldname": "reference_doctype",
			"fieldtype": "Data",
			"width": "250px",
		},
		{
			"fieldname": "attribute",
			"fieldtype": "Data",
			"label": "Attribute",
			"width": "200px",
		},
		{
			"fieldname": "value",
			"label": "Value",
			"fieldtype": "Data",
			"width": "250px",
		},
		{"fieldname": "field", "fieldtype": "Data", "hidden": 1},
		{"fieldname": "name", "fieldtype": "Data", "hidden": 1},
	]


@frappe.whitelist()
def set_value(docname, value):
	frappe.set_value("Specification Value", docname, "value", value)
	frappe.msgprint("Updated", alert=True)

File: inventory_tools/inventory_tools/report/specification/specification.js
// Copyright (c) 2023, AgriTheory and contributors
// For license information, please see license.txt

frappe.query_reports['Specification'] = {
	filters: [
		{
			fieldname: 'specification',
			label: __('Specification'),
			options: 'Specification',
			fieldtype: 'Link',
			reqd: 1,
		},
	],
	refresh: reportview => {
		this.get_datatable_options(frappe.query_report.datatable.options)
	},
	get_datatable_options(options) {
		options.columns[3].editable = true
		return Object.assign(options, {
			getEditor: this.get_editing_object.bind(this),
		})
	},
	get_editing_object(colIndex, rowIndex, value, parent) {
		if (frappe.query_report.datatable.datamanager.data[rowIndex].field != undefined) {
			frappe.show_alert(__('This value cannot be edited in this report'))
			return false
		}
		const control = this.render_editing_input(colIndex, value, parent)
		if (!control) return false
		control.df.change = () => control.set_focus()
		try {
			return {
				initValue: async value => {
					return control.set_value(value)
				},
				setValue: value => {
					let row = frappe.query_report.datatable.datamanager.data[rowIndex]
					let docname = row.name
					if (!value) {
						return control.get_value()
					}
					if (row.indent === 0) {
						return control.get_value()
					}
					if (row.billed_date) {
						return control.get_value()
					}

					return frappe
						.xcall('inventory_tools.inventory_tools.report.specification.specification.set_value', {
							docname: docname,
							value: value,
						})
						.catch(e => {
							return control.get_value(value)
						})
				},
				getValue: async () => {
					return control.get_value()
				},
			}
		} catch (error) {
			console.log(error)
		}
	},
	render_editing_input(colIndex, value, parent) {
		const col = frappe.query_report.datatable.getColumn(colIndex)
		let control = null
		control = frappe.ui.form.make_control({
			df: col,
			parent: parent,
			render_input: true,
		})
		control.set_value(value || '')
		control.toggle_label(false)
		control.toggle_description(false)
		return control
	},
}
